tket/src/Circuit/Multiplexor.cpp

```diff
@@ -0,0 +1,668 @@
+    c.add_box(qcbox, qubits);
+    for (unsigned i : zero_ctrls) {
+      c.add_op<unsigned>(OpType::X, {i});
+    }
+  }
+  return c;
+}
+
+/**
+ * @brief Convert an unsigned to its binary representation
+ * e.g. 4 -> [1, 0, 0]
+ *
+ * @param dec
+ * @param width used to pad zeros
+ * @return std::vector<bool>
+ */
+static std::vector<bool> dec_to_bin(unsigned dec, unsigned width) {
+  auto bs = std::bitset<MAX_N_CONTROLS>(dec);
+  std::vector<bool> bits(width);
+  for (unsigned i = 0; i < width; i++) {
+    bits[width - i - 1] = bs[i];
+  }
+  return bits;
+}
+
+/**
+ * @brief Implement uniformly controlled same-axis rotations (UCR)
+ * with 2^ctrl_qubits SQ rotations, 2^ctrl_qubits CXs, and 2 H gates for X-axis
+ * rotations.
+ *
+ * https://arxiv.org/abs/quant-ph/0410066
+ * This is a special case derived from equation (3)
+ * A UCR gate controlled by n qubits have the decomposition UCR = CX P CX Q
+ * (multiplication order), where P and Q are themselves UCR gates controlled by
+ * n-1 qubits.
+ *
+ * Also notice that CX P CX Q = Q CX P CX, therefore we can control the
+ * direction of each decomposition to avoid adding adjacent CX gates.
+ * e.g. UCR = CX P CX Q = (CX Q' CX P' CX*) CX (CX* P'' CX Q'')
+ * The two CX* can be cancelled,
+ * hence UCR = CX P CX Q = (CX Q' CX P') CX (P''CX Q'')
+ *
+ *
+ * @param angles list of 2^ctrl_qubits angles, angles[i] is the angle activated
+ * by bitstring binary(i)
+ * @param axis can be either Ry or Rz
+ * @param total_qubits the total number of qubits in the final output circuit
+ * @param circ circuit to update
+ * @param direction controls the decomposition direction of each demultiplex
+ * step. Will be implemented as P CX Q if true, and Q CX P if false. If nullopt,
+ * CX P CX Q will be implemented as the root step.
+ *
+ */
+static void recursive_demultiplex_rotation(
+    const std::vector<Expr> &angles, const OpType &axis, unsigned total_qubits,
+    Circuit &circ, std::optional<bool> direction) {
+  unsigned n_rotations = angles.size();
+  unsigned n_qubits = (unsigned)log2(n_rotations) + 1;
+  unsigned mid = (unsigned)(n_rotations / 2);
+  std::vector<Expr> p_angles;
+  std::vector<Expr> q_angles;
+  for (unsigned i = 0; i < mid; i++) {
+    p_angles.push_back((angles[i] - angles[mid + i]) / 2);
+    q_angles.push_back((angles[i] + angles[mid + i]) / 2);
+  }
+  if (direction != std::nullopt && !direction.value()) {
+    std::swap(p_angles, q_angles);
+  }
+  if (q_angles.size() == 1) {
+    circ.add_op<unsigned>(axis, q_angles[0], {total_qubits - 1});
+  } else {
+    recursive_demultiplex_rotation(q_angles, axis, total_qubits, circ, true);
+  }
+  circ.add_op<unsigned>(
+      OpType::CX, {total_qubits - n_qubits, total_qubits - 1});
+  if (p_angles.size() == 1) {
+    circ.add_op<unsigned>(axis, p_angles[0], {total_qubits - 1});
+  } else {
+    recursive_demultiplex_rotation(p_angles, axis, total_qubits, circ, false);
+  }
+  if (direction == std::nullopt) {
```

Comment:
I explained in the function docstring a little. 
true: indicates demultiplexing the left multiplexor with P CX Q (multiplication order)
false: indicates demultiplexing the right multiplexor with Q CX P
notice that two CXs are canceled.

null: indicates demultiplexing the root multiplexor with CX P CX Q

Reply:
I appreciate it's slightly more cumbersome, but I think an ENUM would be better here instead of true/false/NULL. Also means the documentation can be moved to the ENUM class which might present neater.